### PR TITLE
Optimize the PyTorch CUDA implementation for Criss Cross Attention

### PR DESCRIPTION
**Pull request by wondervictor**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
the original CUDA implementation of criss-cross attention processing batches in the kernel, which is a little slow.

## Modification
Increase the `d3` for each thread block to `n*d3` therefore each thread only process one pixel for each image.

```
int d3 = h + w - 1;
dim3 blocks(d1, d2, d3 * n);
```
`n` is the batch size.

*speed test(s)* on one NVIDIA TitanXp, input tensor: `(2, 256, 128, 256)`

|Operator.|Original|Modified|
|:--|:--:|:--:|
|CAWeightFunction Forward|0.1162|0.1051|
|CAWeightFunction Backward|1.8623|1.8148|
|CAMapFunction Forward|0.0734|0.0697|
|CAMapFunction Backward|1.8523|1.8057|

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by wondervictor** [comment]
waiting for response...

**Comment by ZwwWayne** [comment]
lint failed, could you fix the lint issue first?

**Review by grimoire — COMMENTED** [review]
You can add `deviceGuard` in `cc_attention_cuda.cu` to prevent device error when using non default gpu:

```python
at::cuda::CUDAGuard device_guard(input.device());
```

File: mmcv/ops/csrc/cc_attention_cuda_kernel.cuh [hunk 1]
```diff
@@ -44,23 +42,22 @@ __global__ void ca_backward_kernel_t(const T *dw, const T *t, const T *f, T *dt,
...
+    for (int i = 0; i < height; ++i)  {
+      if (i == y) continue;
+      int j = i<y ? i : i-1;
 
-        T _dw = dw[(batch * len + width + j) * sp + y * width + x];
-        T _f = f[(batch * chn + plane) * sp + i * width + x];
-        dt[(batch * chn + plane) * sp + y * width + x] += _dw * _f;
-      }
+      float _dw = dw[(batch * len + width + j) * sp + y*width + x];
+      float _f = f[(batch * chn + plane) * sp + i*width + x];
```
[the remark this answers is not in the document]
Comment (wondervictor) [reply mmcv/ops/csrc/cc_attention_cuda_kernel.cuh:59]: Sorry, I'll replace `float` with `T`.

File: mmcv/ops/csrc/cc_attention_cuda_kernel.cuh [hunk 2]
```diff
@@ -14,24 +14,22 @@ __global__ void ca_forward_kernel(const T *t, const T *f, T *weight, int num,
...
+
+      if (z < width) {
+        int i = z;
+        T _f = f[(batch * chn + plane) * sp + y*width + i];
+        weight[(batch * len + i) * sp + y*width + x] += _t*_f;
+      } else {
+        int i = z - width;
+        int j = i<y ? i : i+1;
+        T _f = f[(batch * chn + plane) * sp + j*width + x];
+        weight[(batch * len + width + i) * sp + y*width + x] += _t*_f;
```
[the remark this answers is not in the document]
Comment (wondervictor) [reply mmcv/ops/csrc/cc_attention_cuda_kernel.cuh:32]: OK, I'll fix it.

File: mmcv/ops/csrc/cc_attention_cuda_kernel.cuh [hunk 3]
```diff
@@ -130,25 +125,23 @@ __global__ void ca_map_backward_kernel_w(const T *dout, const T *weight,
...
+    int gidx = batch * chn * sp;
+    if (z < width) {
+      gidx += y * width + z;
+    } else {
+      int j = z - width;
+      j = j < y ? j : j + 1;
+      gidx += j * width + x;
+    }
+    for(int plane = 0; plane < chn; plane ++){
+      dw[widx] += dout[dout_idx + plane * sp] * g[gidx+plane*sp];
```
[the remark this answers is not in the document]
Comment (wondervictor) [reply mmcv/ops/csrc/cc_attention_cuda_kernel.cuh:144]: Ok, thanks for good advice!

**Comment by ZwwWayne** [comment]
kindly ping @grimoire @wondervictor , we plan to merge this PR in this version, i.e., before this Friday.

**Comment by ZwwWayne** [comment]
Done in https://github.com/open-mmlab/mmcv/pull/1143

**Pull request closed without merging**